Added database server and service support.

aznashwan opened the pull request:
Added support for database server and service CR_D.

This pull depends on: https://github.com/Azure/azure-sdk-for-go/pull/173

ahmetalpbalkan commented:
Feel free to update the SendAzurePostRequest method to return the response []byte for that. :-)

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+package database
```
can we call this `sql` to be consistent with the nuget naming scheme as well https://www.nuget.org/packages/Microsoft.WindowsAzure.Management.Sql/ ? We have a product called DocumentDB as well and that's `database`, too.

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505699.aspx
+func (c DatabaseClient) CreateDatabaseServer(params DatabaseServerCreateParams) error {
+	if params.Location == "" {
```
no need to make these parameter checks unless they're required in constructing the URL, let the API fail with an error.

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+
+// DatabaseClient defines various database CRUD operations.
+// It contains a management.Client for making the actual http calls.
+type DatabaseClient struct {
```
SqlDatabaseClient would be better.

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+	return err
+}
+
+// ListDatabaseServers issues a request for a list of all database servers off Azure.
```
let's be consistent with the docs link below

> ListDatabaseServers retrieves the Azure SQL Database servers for a given subscription.

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// Database represents the set of possible parameters when issuing
+// a database creation to Azure, and reading a list response from Azure.
```
please do not reuse the same type `Database` for both serializing to requests and deserializing from responses. you can alternatively embed this type (make it unexported) into two exported types with xxxRequest and yyyResponse.

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505699.aspx
+type DatabaseServerCreateParams struct {
+	XMLName  xml.Name `xml:"http://schemas.microsoft.com/sqlazure/2010/12/ Server"`
```
is space before `Server` intentional?

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+// https://msdn.microsoft.com/en-us/library/azure/dn505702.aspx
+type DatabaseServer struct {
+	Name     string
+	UserName string `xml:"AdministratorLogin"`
```
use the same name from XML response

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+	Name     string
+	UserName string `xml:"AdministratorLogin"`
+	Location string
+	DNSName  string `xml:"FullyQualifiedDomainName"`
```
use the same name from XML response

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+type ListDatabaseServersResponse struct {
+	XMLName xml.Name         `xml:"Servers"`
```
I believe you don't need XMLName here. The root element of XML, in this case `<Servers>` will be figured out by the xml pkg automatically.

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+// ListDatabaseServers issues a request for a list of all database servers off Azure.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505702.aspx
+func (c DatabaseClient) ListDatabaseServers() ([]DatabaseServer, error) {
```
ListServers, let's be consistent with the API naming

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+// DeleteDatabaseServer issues an Azure database server deletion request.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505695.aspx
+func (c DatabaseClient) DeleteDatabaseServer(name string) error {
```
DeleteServer

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+// CreateDatabaseServer issues a database server creation request to Azure.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505699.aspx
+func (c DatabaseClient) CreateDatabaseServer(params DatabaseServerCreateParams) error {
```
CreateServer

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	err = xml.Unmarshal(data, &resp)
+	return resp.Servers, err
```
Do not return the internal field, just return the response object `ListDatabaseServersResponse` and let the user figure out.

If some other field gets added right next to `.Servers` field, we'll be introducing breaking change to return that as well, we follow returning xxxResponse objects as a convention in the codebase.

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+	ServiceObjectiveId string `xml:",omitempty"`
+}
+
+type ListDatabsesResponse struct {
```
typo

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+
+type ListDatabsesResponse struct {
+	XMLName   xml.Name   `xml:"http://schemas.microsoft.com/windowsazure ServiceResources"`
+	Databases []Database `xml:"http://schemas.microsoft.com/windowsazure ServiceResource"`
```
Use the same name with the API: `ServiceResource`. no need for introducing new names/concepts.

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+	DatabaseServer     string `xml:"-"`
+	Edition            string `xml:",omitempty"`
+	CollationName      string `xml:",omitempty"`
+	MaxSize            int64  `xml:"MaxSizeBytes,omitempty"`
```
use the same name as the API.

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+	Name    string
+	// DatabaseServer should contain the name of the
+	// server the database is/should be hosted on.
+	DatabaseServer     string `xml:"-"`
```
is this `xml:"-"` because this is used as a common type as xxxRequest and yyyResponse objects?

ahmetalpbalkan commented on management/database/entities.go:
```diff
@@ -0,0 +1,67 @@
+// https://msdn.microsoft.com/en-us/library/azure/dn505699.aspx
+type DatabaseServerCreateParams struct {
+	XMLName  xml.Name `xml:"http://schemas.microsoft.com/sqlazure/2010/12/ Server"`
+	UserName string   `xml:"AdministratorLogin"`
```
use the same name as the API (same below)

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+	if params.Name == "" {
+		return fmt.Errorf(errParamNotSpecified, "Name")
+	}
+	if params.DatabaseServer == "" {
```
`Server` definitely should not go into `Database` (which should be `DatabaseCreateParameters`). This signature should read as:

    func (c DatabaseClient) CreateDatabase(server string, params DatabaseCreateParameters)

ahmetalpbalkan commented on management/database/client.go:
```diff
@@ -0,0 +1,154 @@
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505701.aspx
+func (c DatabaseClient) CreateDatabase(params Database) error {
+	if params.Name == "" {
```
do not have logic for checking which request-body parameters are required. let the API do it.

ahmetalpbalkan commented:
@aznashwan apparently the `SendAzurePostRequest` does not return a response body. you can change it to return `([]byte, management.OperationID, error)` if you like.

In the rest of the Azure REST APIs, the POST requests usually take some time and run in the background and expect user to poll the API with `management.OperationID` for operation status, however apparently all SQL Azure APIs are effective immediately when server responds with a result to the REST call...

aznashwan commented:
@ahmetalpbalkan : thanks for the review.

I've addressed all your points and awaiting any further input you guys might have :smile: 

PS: as a side-note; I was in a conflict between your preferring the names used by the API and what golint suggests; I decided to go ahead and rename [this](https://github.com/Azure/azure-sdk-for-go/pull/171/files#diff-fdf8224c399357a019e9146a15741be0R57) field so as to not curse the project with a red CI build forever.

ahmetalpbalkan commented on management/sql/client.go:
```diff
@@ -0,0 +1,163 @@
+// CreateDatabase creates a new Microsoft Azure SQL Database on the given database server.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505701.aspx
+func (c SqlDatabaseClient) CreateDatabase(params DatabaseCreateParams, server string) error {
```
server should come first, params next

ahmetalpbalkan commented on management/sql/client.go:
```diff
@@ -0,0 +1,163 @@
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505701.aspx
+func (c SqlDatabaseClient) CreateDatabase(params DatabaseCreateParams, server string) error {
+	if params.Name == "" {
```
do not check for `params.name`, it's not required to make the request, it's rather required to do the operation successfully and that's what API server checks in the input as well. so leave the check to the API server when you can.

ahmetalpbalkan commented on management/sql/client.go:
```diff
@@ -0,0 +1,163 @@
+// GetDatabase gets the details for an Azure SQL Database.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505708.aspx
+func (c SqlDatabaseClient) GetDatabase(database, server string) (ServiceResource, error) {
```
`server ` first.

ahmetalpbalkan commented on management/sql/client.go:
```diff
@@ -0,0 +1,163 @@
+// DeleteDatabase deletes the Azure SQL Database off the given server.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505705.aspx
+func (c SqlDatabaseClient) DeleteDatabase(database, server string) error {
```
`server` first.

ahmetalpbalkan commented on management/sql/client.go:
```diff
@@ -0,0 +1,163 @@
+// ListServers retrieves the Azure SQL Database servers for this subscription.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505702.aspx
+func (c SqlDatabaseClient) ListServers() (ListDatabaseServersResponse, error) {
```
should be `ListServersResponse`

ahmetalpbalkan commented on management/sql/entities.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+type ListDatabaseServersResponse struct {
+	DatabaseServers []DatabaseServer `xml:"Server"`
```
"Server" should be "ServiceResource" otherwise it will not work:

http://play.golang.org/p/4_tpfVF6dK it doesn't match the response.

ahmetalpbalkan replied:
Hmm actually many of the fields in DatabaseServer does not fit what's returned from the API?

aznashwan replied:
I don't use ?countview=generic as it offers a lot more information that really needed (in particular stuff like Type and the Links).
Without that param; the response looks like the second example from that page; which is the one I'm targeting.

ahmetalpbalkan commented:
@aznashwan thanks a lot for the updates, I just left a few more comments. The Id→ID conversion is fine, we do it in other places as well. :-)

aznashwan commented:
@ahmetalpbalkan : thanks a lot for the time!
I've done the necessary changes for both this and the POST response pull and and looking forward to seeing the code in and usable :smile:

ahmetalpbalkan commented on management/sql/entities.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// ServiceResource represents the set of parameters obtained from a database
+// get or lost call.
```
`list`

aznashwan replied:
I'm sorry; but I don't fully get what you mean here.

aznashwan replied:
sorry about that; I noticed the type now

ahmetalpbalkan commented on management/sql/entities.go:
```diff
@@ -0,0 +1,78 @@
+// ServiceResource represents the set of parameters obtained from a database
+// get or lost call.
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505701.aspx
```
Link points to `Create Database`. :)

ahmetalpbalkan commented on management/sql/entities.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+type ListDatabasesResponse struct {
+	XMLName          xml.Name          `xml:"http://schemas.microsoft.com/windowsazure ServiceResources"`
```
All these schema URIs and xml.Name(s) are redundant. We prefer to not to return XMLName to the user where possible:  http://play.golang.org/p/Za2dGTHf3G

This struct probably should just be like:

```go
type ListDatabasesResponse struct {
	ServiceResources []ServiceResource `xml:"ServiceResource"`
}
```

aznashwan replied:
Thanks for the example. Fixed it now.

ahmetalpbalkan commented on management/sql/entities.go:
```diff
@@ -0,0 +1,78 @@
+//
+// https://msdn.microsoft.com/en-us/library/azure/dn505701.aspx
+type ServiceResource struct {
+	XMLName            xml.Name `xml:"http://schemas.microsoft.com/windowsazure ServiceResource"`
```
XMLName field redundant.

ahmetalpbalkan commented:
Looks great @aznashwan. Did you have a chance to test each of these calls? I'm just verbally asking because sometimes API doesn't work as it's documented (e.g. Ordering of request fields sometimes matter). Does Terraform use this to verify functionality of this package?

aznashwan commented:
@ahmetalpbalkan: I will get to actually writing the resources for Terraform which use this portion of the SDK in a moment.
I have tested each method "manually" (I can share the code snippets if you like), even after all the changes, and I can vouch they work.

ahmetalpbalkan commented:
Awesome, thanks.
